### PR TITLE
first attempt of moving to hooks

### PR DESCRIPTION
**Pull request by costinEEST**
It is a work in progress. 
Still have to figure it out how to extract `this.setState()` from the if conditions..
Also, some props are not declared initially but are used inside the main component.

### CODE REVIEW

**Comment by baconcheese113** [comment]
@ccampean If you're trying to do something when a variable is changed, React.useEffect() should work. Another option might be using the function version of setState 
```setState((currentState) => {return newState})```

File: MultiSlider.js [hunk 1]
```diff
@@ -1,573 +1,486 @@
...
   Platform,
-  Dimensions,
   I18nManager,
   ImageBackground,
 } from 'react-native';
 
-import DefaultMarker from './DefaultMarker';
-import DefaultLabel from './DefaultLabel';
+import { DefaultMarker } from './DefaultMarker';
+import { DefaultLabel } from './DefaultLabel';
```
Comment (sibelius) [inline MultiSlider.js]: not sure we need to change this

File: MultiSlider.js [hunk 2]
```diff
@@ -1,573 +1,486 @@
...
+  onToggleOne = () => {},
+  onToggleTwo = () => {},
+  enabledOne = true,
+  enabledTwo = true,
+  allowOverlap = false,
+  snapped = false,
+  vertical = false,
+  minMarkerOverlapDistance = 0,
+}) {
+  let [onePressed, setOnePressed] = useState(false);
```
Comment (sibelius) [inline MultiSlider.js]: prefer const here

File: MultiSlider.js [hunk 3]
```diff
@@ -1,573 +1,486 @@
...
+  let [twoPressed, setTwoPressed] = useState(false);
+  let [valueOne, setValueOne] = useState(values[0]);
+  let [valueTwo, setValueTwo] = useState(values[1]);
+  let [pastOne, setPastOne] = useState(initialValues[0]);
+  let [pastTwo, setPastTwo] = useState(initialValues[1]);
+  let [positionOne, setPositionOne] = useState(initialValues[0]);
```
Comment (sibelius) [inline MultiSlider.js]: this should be after initialValues calculation

File: MultiSlider.js [hunk 4]
```diff
@@ -1,573 +1,486 @@
...
+  let startTwo = () => {
+    if (enabledTwo) {
+      onValuesChangeStart();
+      setTwoPressed(!twoPressed);
     }
   };
 
-  moveOne = gestureState => {
-    if (!this.props.enabledOne) {
+  let moveOne = gestureState => {
```
Comment (sibelius) [inline MultiSlider.js]: const

File: MultiSlider.js [hunk 5]
```diff
@@ -1,573 +1,486 @@
...
+  let [pastTwo, setPastTwo] = useState(initialValues[1]);
+  let [positionOne, setPositionOne] = useState(initialValues[0]);
+  let [positionTwo, setPositionTwo] = useState(initialValues[1]);
+
+  let optionsArray = optionsArray || createArray(min, max, step);
+  let stepLength = sliderLength / optionsArray.length;
+  let initialValues = values.map(value =>
+    valueToPosition(value, optionsArray, sliderLength),
+  );
+  let customPanResponder = (start, move, end) =>
```
Comment (sibelius) [inline MultiSlider.js]: should this be inside `subscribePanResponder`?

File: MultiSlider.js [hunk 6]
```diff
@@ -1,573 +1,486 @@
...
             }
-            this.props.onValuesChange(change);
+            onValuesChange(change);
 
-            this.props.onMarkersPosition([
-              this.state.positionOne,
-              this.state.positionTwo,
-            ]);
+            onMarkersPosition([positionOne, positionTwo]);
           },
```
Comment (sibelius) [inline MultiSlider.js]:
```jsx
setValueOne(value);
let change = [value];
if (valueTwo) {
   change.push(valueTwo)
}
onValuesChange(change);
onMarkersPosition([positionOne, positionTwo]);
```

File: MultiSlider.js [hunk 7]
```diff
@@ -1,573 +1,486 @@
...
       },
     );
   };
 
-  componentDidUpdate(prevProps, prevState) {
-    const {
-      positionOne: prevPositionOne,
-      positionTwo: prevPositionTwo,
-    } = prevState;
+  let _panResponderBetween = customPanResponder(
```
Comment (sibelius) [inline MultiSlider.js]: useRef

File: MultiSlider.js [hunk 8]
```diff
@@ -1,573 +1,486 @@
...
+    gestureState => {
+      endOne(gestureState);
+      endTwo(gestureState);
+    },
+  );
+
+  let _panResponderOne = customPanResponder(startOne, moveOne, endOne);
 
-    const { positionOne, positionTwo } = this.state;
+  let _panResponderTwo = customPanResponder(startTwo, moveTwo, endTwo);
```
Comment (sibelius) [inline MultiSlider.js]: useRef

**Comment by ptomasroos** [comment]
Great that you’re taking a attempt on this! Do not be afraid of making some breaking changes if needed. Could be released as a new major. We could probably simplify if needed a bit by making it a controlled component by not keeping both local and state and props being passed in. 

That’s at least my feeling the most “broken” part of today. That the internal state must be resetted when a parent updates the properties. 

@sibelius thanks for helping out!

**Comment by sibelius** [comment]
for controlled/uncontrolled you can use this hook

https://gist.github.com/sibelius/d44c23c1b50f2edb393718b0236dbb20

`useControlledState`

File: MultiSlider.js (hunk 1 above)
Comment (ccampean) [reply MultiSlider.js]: They were rewritten as named exports.

File: MultiSlider.js (hunk 2 above)
Comment (ccampean) [reply MultiSlider.js]: Done.

File: MultiSlider.js (hunk 3 above)
Comment (ccampean) [reply MultiSlider.js]: Done.

File: MultiSlider.js (hunk 4 above)
Comment (ccampean) [reply MultiSlider.js]: Applied `const` everywhere was possible.

File: MultiSlider.js (hunk 5 above)
Comment (ccampean) [reply MultiSlider.js]: Did it.

File: MultiSlider.js (hunk 6 above)
Comment (ccampean) [reply MultiSlider.js]: First, I have to extract somehow setting the state from the `if` statement or to break somehow the `moveOne` function.

File: MultiSlider.js (hunk 7 above)
Comment (ccampean) [reply MultiSlider.js]: I want to make the component uncontrolled, but what do I have to remove/rewrite?

File: MultiSlider.js (hunk 8 above)
Comment (ccampean) [reply MultiSlider.js]: I want to make the component uncontrolled, but what do I have to remove/rewrite?

**Comment by ccampean** [comment]
Thanks all for your support.
@sibelius : can you give me more hints, of how to make the components uncontrolled by React. Also, I still don't know how to break the functions that are setting the state inside `if` conditions.

File: MultiSlider.js [hunk 9]
```diff
@@ -59,34 +59,22 @@ function MultiSlider({
...
-  const optionsArray = optionsArray || createArray(min, max, step);
-  const stepLength = sliderLength / optionsArray.length;
-  const initialValues = values.map(value =>
-    valueToPosition(value, optionsArray, sliderLength)
-  );
+  let optionsArray;
+  let stepLength;
+  let initialValues;
+
+  let _panResponderBetween;
```
Comment (sibelius) [inline MultiSlider.js]: this should be a `useRef`

File: index.d.ts [hunk 10]
```diff
@@ -46,7 +46,7 @@ export interface MultiSliderProps {
     customMarkerRight?: React.ComponentType<MarkerProps>;
     customLabel?: React.ComponentType<LabelProps>;
 
-    isMarkersSeparated?: boolean;
+    areMarkersSeparated?: boolean;
```
Comment (sibelius) [inline index.d.ts]: avoid modifying props, this would be a breaking change

**Pull request closed without merging**